subprojects/core/src/main/java/org/gradle/caching/internal/tasks/BuildCacheKeyInputs.java

```diff
@@ -37,7 +36,14 @@
     private final ImmutableSortedMap<String, HashCode> inputHashes;
     private final ImmutableSortedSet<String> outputPropertyNames;
 
-    public BuildCacheKeyInputs(String taskClass, HashCode classLoaderHash, List<HashCode> actionClassLoaderHashes, ImmutableList<String> actionClassNames, ImmutableSortedMap<String, HashCode> inputHashes, ImmutableSortedSet<String> outputPropertyNames) {
+    public BuildCacheKeyInputs(
+        @Nullable String taskClass,
```

Comment:
This is used as a “null object” already, when a task has no outputs. Having unclear null semantics is a recipe for disaster, as evidenced by the fact that we did not realise that these objects can be null under some circumstances.